client/client.go

```diff
@@ -0,0 +1,462 @@
+	Started() <-chan struct{}
+}
+
+// New returns an implemenetation of DynamicWatcher that is ready to be started with the Start method. An error is
+// returned if Kubernetes clients can't be instantiated with the input Kubernetes configuration.
+func New(config *rest.Config, reconciler Reconciler, options *Options) (DynamicWatcher, error) {
+	if options == nil {
+		options = &Options{}
+	}
+
+	rateLimiter := options.RateLimiter
+	if rateLimiter == nil {
+		rateLimiter = workqueue.DefaultControllerRateLimiter()
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize a Kubernetes client: %w", err)
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize a dynamic Kubernetes client: %w", err)
+	}
+
+	watcher := dynamicWatcher{
+		client:            client,
+		dynamicClient:     dynamicClient,
+		rateLimiter:       rateLimiter,
+		Reconciler:        reconciler,
+		startedChan:       make(chan struct{}),
+		watchedToWatchers: map[ObjectIdentifier]map[ObjectIdentifier]bool{},
+		watcherToWatches:  map[ObjectIdentifier]map[ObjectIdentifier]bool{},
+		watches:           map[ObjectIdentifier]watch.Interface{},
+	}
+
+	return &watcher, nil
+}
+
+// dynamicWatcher implements the DynamicWatcher interface.
+type dynamicWatcher struct {
+	// dynamicClient is a client-go dynamic client used for the dynamic watches.
+	dynamicClient dynamic.Interface
+	// client is a client-go client used for API discovery.
+	client kubernetes.Interface
+	lock   sync.RWMutex
+	Queue  workqueue.RateLimitingInterface
+	Reconciler
+	rateLimiter ratelimiter.RateLimiter
+	// restMapper is used to map ObjectIdentifer objects to an API resource.
+	restMapper apimachinerymeta.RESTMapper
+	// started gets set as part of the Start method.
+	started bool
+	// startedChan is closed when the dynamicWatcher is started. This is exposed to the user through the Started method.
+	startedChan chan struct{}
+	// watchedToWatchers is a map where the keys are ObjectIdentifier objects representing the watched objects.
+	// Each value acts as a set of ObjectIdentifier objects representing the watcher objects.
+	watchedToWatchers map[ObjectIdentifier]map[ObjectIdentifier]bool
+	// watcherToWatches is a map where the keys are ObjectIdentifier objects representing the watcher objects.
+	// Each value acts as a set of ObjectIdentifier objects representing the watched objects.
+	watcherToWatches map[ObjectIdentifier]map[ObjectIdentifier]bool
+	// watches is a map where the keys are ObjectIdentifier objects representing the watched objects and the values
+	// are objects representing the Kubernetes watch API requests.
+	watches map[ObjectIdentifier]watch.Interface
+}
+
+// Start will start the dynamicWatcher and block until the input context is canceled.
+func (d *dynamicWatcher) Start(ctx context.Context) error {
```

Comment:
Nit/opinion: `Run` might be a more accurate name for this? Since this function goes until it stops (and handles some stop/cleanup logic); if I hear about a `Start` function I first assume it will start something asynchronously in the background.

But if this matches a convention used in other places, then totally disregard this comment :)

Reply:
I agree but I chose the name because of the `Manager.Start` method from controller-runtime:
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#example-Manager-Start